src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HealthHeartbeatProvider.cs

```diff
@@ -0,0 +1,226 @@
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility;
+
+    /// <summary>
+    /// Implementation of health heartbeat functionality.
+    /// </summary>
+    internal class HealthHeartbeatProvider : IDisposable, IHeartbeatProvider
+    {
+        /// <summary>
+        /// The default interval between heartbeats if not specified by the user
+        /// </summary>
+        public static int DefaultHeartbeatIntervalMs = 5000;
+
+        /// <summary>
+        /// The default fields to include in every heartbeat sent. Note that setting the value to '*' includes all default fields.
+        /// </summary>
+        public static string DefaultAllowedFieldsInHeartbeatPayload = "*";
+
+        /// <summary>
+        /// The name of the health heartbeat metric item and operation context.
+        /// </summary>
+        public static string HeartbeatSyntheticMetricName = "SDKHeartbeat";
+
+        /// <summary>
+        /// The payload items to send out with each health heartbeat.
+        /// </summary>
+        private Dictionary<string, IHealthHeartbeatPayloadExtension> payloadItems;
+        
+        private bool disposedValue = false; // To detect redundant calls to dispose
+        private int intervalBetweenHeartbeatsMs; // time between heartbeats emitted specified in milliseconds
+        private string enabledHeartbeatPayloadFields; // string containing fields that are enabled in the payload. * means everything available.
+        private TelemetryClient telemetryClient; // client to use in sending our heartbeat
+        private int heartbeatsSent;
+
+        public HealthHeartbeatProvider() : this(DefaultHeartbeatIntervalMs, DefaultAllowedFieldsInHeartbeatPayload)
+        {
+        }
+
+        public HealthHeartbeatProvider(int delayMs) : this(delayMs, DefaultAllowedFieldsInHeartbeatPayload)
+        {
+        }
+
+        public HealthHeartbeatProvider(string allowedPayloadFields) : this(DefaultHeartbeatIntervalMs, allowedPayloadFields)
+        {
+        }
+
+        public HealthHeartbeatProvider(int delayMs, string allowedPayloadFields)
+        {
+            this.enabledHeartbeatPayloadFields = allowedPayloadFields;
+            this.intervalBetweenHeartbeatsMs = delayMs;
+            this.payloadItems = new Dictionary<string, IHealthHeartbeatPayloadExtension>();
+            this.heartbeatsSent = 0; // count up from construction time
+        }
+
+        public int HeartbeatIntervalMs => this.intervalBetweenHeartbeatsMs;
+
+        public string EnabledPayloadFields => this.enabledHeartbeatPayloadFields;
+
+        public bool Initialize(TelemetryConfiguration configuration, int? delayMs = null, string allowedPayloadFields = null)
+        {
+            if (this.telemetryClient == null)
+            {
+                this.telemetryClient = new TelemetryClient(configuration);
```

Comment:
Diagnostics module allows to override the instrumentation key. Consider doing the same here

Reply:
Will follow up with you on this requirement - does overriding the instrumentation key mean re-creating the telemetry client?